### PR TITLE
Bp 51/chore/add impala

### PR DESCRIPTION
**Pull request by mayankgituni**

### CODE REVIEW

File: docker-compose.yml [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+          memory: "4096m"
+
+# Volume to store Log files
+volumes:
+  preprocess-data:
+    driver: local
+
+# private Bridge Network
+networks:
+  impala-net:
```
Comment (omjadas) [inline docker-compose.yml:41]:
```suggestion
```

docker compose creates a default network so this shouldn't be necessary

File: src/main/java/org/apromore/plugin/services/impl/ImpalaJdbcAdaptor.java [hunk 2]
```diff
@@ -0,0 +1,149 @@
...
+     * Execute SQL query on the impala tables.
+     *
+     * @param sqlStatement Sql string
+     * @return List of the result rows
+     */
+    public List<List<String>> executeQuery(String sqlStatement) {
+
+        List<List<String>> resultList = new ArrayList<>();
+
+        try {
```
Comment (omjadas) [inline src/main/java/org/apromore/plugin/services/impl/ImpalaJdbcAdaptor.java:75]: you should use a try with resources

File: src/main/java/org/apromore/plugin/services/impl/TransactionImp.java [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+
+import org.apromore.plugin.services.Transaction;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.stereotype.Service;
+
+/**
+ * Transaction service for executing statements for impala.
+ */
+@Service("transactionService")
+public class TransactionImp implements Transaction {
```
Comment (omjadas) [inline src/main/java/org/apromore/plugin/services/impl/TransactionImp.java:13]:
```suggestion
public class TransactionImpl implements Transaction {
```

File: src/main/java/org/apromore/plugin/services/impl/TransactionImp.java [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+
+    /**
+     * Add the file to the Impala and get a snippet.
+     *
+     * @param fileName File Name
+     * @param limit  Limit of the rows
+     * @return return the snippet of the table.
+     */
+    @Override
+    public List<List<String>> addTableGetSnippet(String fileName, int limit) {
```
Comment (omjadas) [inline src/main/java/org/apromore/plugin/services/impl/TransactionImp.java:26]: I think this should just add the table, if we need a snippet we can call getSnippet afterwards

File: src/main/java/org/apromore/plugin/services/impl/FileHandlerImpl.java [hunk 5]
```diff
@@ -17,17 +19,15 @@
...
 
     /**
      * Create a directory to save the output files to.
      */
     private void generateDirectory() {
         String temporalDir = new File("").getAbsolutePath();
-        File directory = new File(temporalDir + "/files/");
-        this.tempDir = temporalDir + "/files/";
+        File directory = new File(temporalDir + "/preprocess_data/");
+        this.tempDir = temporalDir + "/preprocess_data/";
```
Comment (omjadas) [inline src/main/java/org/apromore/plugin/services/impl/FileHandlerImpl.java:30]: the path to where the files are stored should be an environment variable

File: src/main/java/org/apromore/plugin/services/impl/ImpalaJdbcAdaptor.java [hunk 6]
```diff
@@ -0,0 +1,149 @@
...
+        // Build String
+        String query = "CREATE EXTERNAL TABLE %s " +
+                        "LIKE PARQUET '/preprocess_data/%s' " +
+                        "STORED AS PARQUET " +
+                        "LOCATION '/preprocess_data'";
+
+        query = String.format(query, tableName, fileName);
+        boolean status = false;
+
+        try {
```
Comment (omjadas) [inline src/main/java/org/apromore/plugin/services/impl/ImpalaJdbcAdaptor.java:40]: this should be a try with resources

File: src/main/java/org/apromore/plugin/FileUploadViewModel.java [hunk 7]
```diff
@@ -20,15 +21,41 @@
...
+
+    }
+
+    /**
+     * Convert the list result into String.
+     *
+     * @param tableVal snippet list values
+     * @return converted string
+     */
+    private String createTableOutput(List<List<String>> tableVal) {
```
Comment (omjadas) [inline src/main/java/org/apromore/plugin/FileUploadViewModel.java:46]: is this useful?

File: src/main/java/org/apromore/plugin/services/impl/ImpalaJdbcAdaptor.java [hunk 8]
```diff
@@ -0,0 +1,149 @@
...
+
+        } catch (SQLException e) {
+            System.out.println("Failed to add Table!!");
+            e.printStackTrace();
+        } catch (Exception e) {
+            System.out.println("Failed to add Table!!");
+            e.printStackTrace();
+        } catch (Throwable throwable) {
+            throwable.printStackTrace();
+        }
```
Comment (omjadas) [inline src/main/java/org/apromore/plugin/services/impl/ImpalaJdbcAdaptor.java:61]:
```suggestion
        } catch (SQLException e) {
            System.out.println("Failed to add Table!!");
            e.printStackTrace();
        }
```

File: src/main/java/org/apromore/plugin/services/impl/ImpalaJdbcAdaptor.java [hunk 9]
```diff
@@ -0,0 +1,149 @@
...
+        } catch (SQLException e) {
+            e.printStackTrace();
+        }
+    }
+
+    /**
+     * Exit the JDBC connection.
+     * @throws Throwable Close connection error.
+     */
+    private void exitConnection() throws Throwable {
```
Comment (omjadas) [inline src/main/java/org/apromore/plugin/services/impl/ImpalaJdbcAdaptor.java:146]:
```suggestion
    private void exitConnection() throws SQLException {
```

Comment (omjadas) [reply src/main/java/org/apromore/plugin/services/impl/ImpalaJdbcAdaptor.java:146]: update the javadoc as well

File: src/main/java/org/apromore/plugin/services/impl/ImpalaJdbcAdaptor.java [hunk 10]
```diff
@@ -0,0 +1,149 @@
...
+            System.out.println("Executed: " + sqlStatement);
+            exitConnection();
+
+        } catch (SQLException e) {
+            e.printStackTrace();
+        } catch (Exception e) {
+            e.printStackTrace();
+        } catch (Throwable throwable) {
+            throwable.printStackTrace();
+        }
```
Comment (omjadas) [inline src/main/java/org/apromore/plugin/services/impl/ImpalaJdbcAdaptor.java:120]:
```suggestion
        } catch (SQLException e) {
            e.printStackTrace();
        }
```

File: Dockerfile [hunk 11]
```diff
@@ -10,10 +11,16 @@ COPY src ./src
...
 
+# Change Volume permissions from root to jetty
+RUN mkdir /var/lib/jetty/preprocess_data
+RUN chown jetty:jetty /var/lib/jetty/preprocess_data
+USER jetty
+
 EXPOSE 8080
 
-CMD java -jar $JETTY_HOME/start.jar
+CMD java -jar $JETTY_HOME/start.jar
```
[the remark this answers is not in the document]
Comment (omjadas) [reply Dockerfile:26]: there should be trailing newlines for all files

File: src/main/java/org/apromore/plugin/services/impl/ImpalaJdbcAdaptor.java [hunk 12]
```diff
@@ -0,0 +1,149 @@
...
+     */
+    public boolean addTable(String tableName, String fileName) {
+
+        String sqlStatementDrop = "DROP TABLE IF EXISTS " + tableName;
+
+        // Build String
+        String query = "CREATE EXTERNAL TABLE %s " +
+                        "LIKE PARQUET '/preprocess_data/%s' " +
+                        "STORED AS PARQUET " +
+                        "LOCATION '/preprocess_data'";
```
Comment (omjadas) [inline src/main/java/org/apromore/plugin/services/impl/ImpalaJdbcAdaptor.java:35]:
```suggestion
        String query = "CREATE EXTERNAL TABLE %s " +
                "LIKE PARQUET '/preprocess_data/%s' " +
                "STORED AS PARQUET " +
                "LOCATION '/preprocess_data'";
```

File: src/main/java/org/apromore/plugin/services/impl/FileHandlerImpl.java [hunk 13]
```diff
@@ -91,4 +102,30 @@ public String writeFiles(Media media) {
...
+     *
+     * @param filePath Path of the file in the volume.
+     * @throws Exception for the file permission change failure.
+     */
+    private void changeFilePermission(String filePath) throws Exception {
+        Path path = Paths.get(filePath);
+        Set<PosixFilePermission> permissions = EnumSet.of(OWNER_READ,
+                                                          OWNER_WRITE,
+                                                          GROUP_READ,
+                                                          OTHERS_READ);
```
Comment (omjadas) [inline src/main/java/org/apromore/plugin/services/impl/FileHandlerImpl.java:118]:
```suggestion
        Set<PosixFilePermission> permissions = EnumSet.of(
            OWNER_READ,
            OWNER_WRITE,
            GROUP_READ,
            OTHERS_READ);
```

File: src/main/java/org/apromore/plugin/services/impl/ImpalaJdbcAdaptor.java [hunk 14]
```diff
@@ -0,0 +1,149 @@
...
+
+        return resultList;
+    }
+
+    /**
+     * Constructing the JDBC connection.
+     */
+    private void initConnection() {
+        try {
+            //
```
Comment (omjadas) [inline src/main/java/org/apromore/plugin/services/impl/ImpalaJdbcAdaptor.java:130]:
```suggestion
```

File: src/main/java/org/apromore/plugin/services/impl/ImpalaJdbcAdaptor.java [hunk 15]
```diff
@@ -0,0 +1,149 @@
...
+
+import org.springframework.stereotype.Component;
+
+/**
+ * Impala JDBC Adaptor class to connect and execute queries.
+ */
+@Component
+public class ImpalaJdbcAdaptor {
+    // Impala connection info
+    private final String connectionUrl = "jdbc:impala://apacheimpala:21050";
```
[the remark this answers is not in the document]
Comment (mayankgituni) [reply src/main/java/org/apromore/plugin/services/impl/ImpalaJdbcAdaptor.java:15]: Sure, Lets setup dev and test environment variables.

File: src/main/java/org/apromore/plugin/services/impl/ImpalaJdbcAdaptor.java (hunk 6 above)
Comment (mayankgituni) [reply src/main/java/org/apromore/plugin/services/impl/ImpalaJdbcAdaptor.java:40]: Not sure what it means.

File: src/main/java/org/apromore/plugin/FileUploadViewModel.java (hunk 7 above)
Comment (mayankgituni) [reply src/main/java/org/apromore/plugin/FileUploadViewModel.java:46]: Maybe not in a long run, but it was there to show the table can be printed.

File: src/main/java/org/apromore/plugin/services/impl/ImpalaJdbcAdaptor.java (hunk 15 above)
[the remark this answers is not in the document]
Comment (omjadas) [reply src/main/java/org/apromore/plugin/services/impl/ImpalaJdbcAdaptor.java:15]: it just needs to be one environment variable

File: src/main/java/org/apromore/plugin/services/impl/ImpalaJdbcAdaptor.java (hunk 6 above)
Comment (omjadas) [reply src/main/java/org/apromore/plugin/services/impl/ImpalaJdbcAdaptor.java:40]: https://docs.oracle.com/javase/tutorial/essential/exceptions/tryResourceClose.html

**Pull request merged**